### PR TITLE
Add moondream2

### PR DESCRIPTION
**Pull request by KissrainKlein**
本 PR 完成 Moondream2 从参考实现到 PaddleFormers 的迁移，包含模型组网、图像处理与 Processor、Auto 系列注册、前向 logits 精度验证、前 10 个生成 token 对齐、PaddleFormers 与 ms-swift 的 300-step 训练 loss 曲线对比，以及模型单测验证。

### 1. 模型实现与注册

新增 Moondream2 模型、配置、图像处理器和 Processor：

- `paddleformers/transformers/moondream2/configuration.py`
- `paddleformers/transformers/moondream2/modeling.py`
- `paddleformers/transformers/moondream2/image_processor.py`
- `paddleformers/transformers/moondream2/processor.py`
- `paddleformers/transformers/moondream2/__init__.py`

完成以下 PaddleFormers 接入：

- 支持 `moondream2` 配置识别；
- 注册 `Moondream2Config`；
- 注册 `Moondream2ForConditionalGeneration`；
- 注册 `Moondream2ImageProcessor`；
- 注册 `Moondream2Processor`；
- 完成 `paddleformers.transformers` 顶层公共 API 导出。

模型实现包括：

- 全局图像 crop 和重叠局部 crops；
- ViT patch embedding、attention、MLP 和视觉投影；
- 局部 crop 特征重建与固定 729 个图像 token 投影；
- Grouped Query Attention；
- partial RoPE；
- parallel pre-norm residual block；
- BOS 与图像特征组成的 730-token 双向可见图像前缀；
- 纯文本和图文多模态前向；
- KV cache 和增量解码；
- 多模态 causal language modeling loss；
- `GenerationMixin` 自回归生成；
- caption、query、detect、point 任务 prompt；
- 坐标与尺寸的 Fourier 特征编码和 region 编解码基础组件；
- 视觉 Transformer 训练路径 recompute。

### 2. 前向 logits 精度验证

验证对象为 Moondream2 参考实现与 PaddleFormers 实现，共使用 7 个 case：

- 2 个图像问答 case；
- 5 个中文纯文本 case，覆盖数学、物理、历史、生物和 Python 编程。

在 FP32 logits 输出接口下，Paddle 模型内部使用 FP64 推理，最终 logits 转回 FP32。使用以下精度标准：

- `atol=1e-2`
- `rtol=0`
- 对完整序列的全部 logits 执行逐元素比较。

7/7 个 case 均通过精度验证：

| Case | Logits shape | 最大绝对误差 | 平均绝对误差 |
|---|---:|---:|---:|
| image-1 | `[1, 754, 51200]` | `2.5606155e-3` | `1.7744377e-4` |
| image-2 | `[1, 754, 51200]` | `3.5009384e-3` | `2.0670512e-4` |
| text-math | `[1, 68, 51200]` | `6.7710876e-5` | `4.8662073e-6` |
| text-physics | `[1, 42, 51200]` | `7.2479248e-5` | `5.0301455e-6` |
| text-history | `[1, 34, 51200]` | `6.4849854e-5` | `4.8620745e-6` |
| text-biology | `[1, 26, 51200]` | `4.6968460e-5` | `4.3805003e-6` |
| text-python | `[1, 51, 51200]` | `6.8664551e-5` | `5.6967506e-6` |

验证结果：

- 最大 logits 绝对误差：`3.5009384e-3`
- 7 个 case 的完整序列 Top-1 一致率均为 `1.0`
- 7 个 case 的最后位置 Top-1 均一致
- HF/Paddle 非有限 logits 数均为 0

说明：以上结果对应 FP32 输出接口，Paddle 模型内部参数和计算使用 FP64，以规避原生 Paddle FP32 算术累积误差；不将该结果表述为原生 Paddle FP32 GEMM 与 PyTorch FP32 GEMM 的严格对齐。

### 3. 生成前向进度对齐

在相同权重、greedy decoding 下，对同一组 7 个 case 验证生成 token：

- 图像 case 使用双方一致的 `full-forward-no-cache` 路径；
- 文本 case 使用增量 KV cache；
- 比较前 10 个生成 token IDs；
- 如果双方共同提前生成 EOS，则比较完整短序列。

7/7 个 case 均通过：

#### image-1

```text
HF first 10 new token IDs:
[826, 1985, 23231, 264, 3148, 41829, 3797, 298, 42911, 51022]

Paddle first 10 new token IDs:
[826, 1985, 23231, 264, 3148, 41829, 3797, 298, 42911, 51022]
```

#### image-2

```text
HF first 10 new token IDs:
[35581, 252, 32487, 259, 23695, 143, 29595, 234, 39635, 25387]

Paddle first 10 new token IDs:
[35581, 252, 32487, 259, 23695, 143, 29595, 234, 39635, 25387]
```

#### text-math

```text
HF first 10 new token IDs:
[2586, 42934, 43619, 41353, 916, 7951, 41278, 41918, 5882, 41203]

Paddle first 10 new token IDs:
[2586, 42934, 43619, 41353, 916, 7951, 41278, 41918, 5882, 41203]
```

#### text-physics

```text
HF generated token IDs:
[489, 3211, 286, 15049, 41649, 1519, 13051, 34]

Paddle generated token IDs:
[489, 3211, 286, 15049, 41649, 1519, 13051, 34]
```

双方均在第 8 个 token 后生成 EOS，完整短序列一致。

#### text-history

```text
HF first 10 new token IDs:
[489, 19558, 49746, 41183, 5347, 41193, 5247, 47141, 433, 49437]

Paddle first 10 new token IDs:
[489, 19558, 49746, 41183, 5347, 41193, 5247, 47141, 433, 49437]
```

#### text-biology

```text
HF first 10 new token IDs:
[16747, 41216, 7777, 4808, 749, 344, 1558, 267, 5459, 41444]

Paddle first 10 new token IDs:
[16747, 41216, 7777, 4808, 749, 344, 1558, 267, 5459, 41444]
```

#### text-python

```text
HF first 10 new token IDs:
[2586, 42934, 1736, 41197, 706, 344, 45146, 1763, 225, 38]

Paddle first 10 new token IDs:
[2586, 42934, 1736, 41197, 706, 344, 45146, 1763, 225, 38]
```

最终结果：

- 7 个 case 的前 10 个生成 token IDs 完全一致；
- `text-physics` 双方共同提前 EOS，生成的 8 个 token IDs 完全一致；
- 所有 case 的 `first_difference_index` 均为空。

### 4. PaddleFormers 与 ms-swift 训练 loss 曲线对比

使用相同数据、训练顺序和超参数，对 PaddleFormers 与 ms-swift 进行 FP32 全参数 SFT 对比。

训练配置：

- 训练数据：250 条；
- 数据组成：240 条文本样本和 10 条图像样本；
- 验证数据：22 条；
- 每设备 batch size：`1`；
- gradient accumulation：`1`；
- optimizer steps：`300`；
- 初始学习率：`1e-5`；
- scheduler：linear；
- warmup：`20` steps；
- weight decay：`0`；
- max grad norm：`1.0`；
- max sequence length：`1024`；
- seed/data seed：`23`；
- 数据顺序：固定顺序，不进行 shuffle；
- 精度：FP32；
- 训练方式：full finetuning。

对齐结果：

| 指标 | PaddleFormers | ms-swift / 对比结果 |
|---|---:|---:|
| 有效 step 数 | `300` | `300` |
| Step 1 loss | `7.79752398` | `7.79752350` |
| Step 300 loss | `8.32802391` | `8.02783012` |
| 300-step 平均 loss | `5.04718745` | `5.04272918` |
| 最小 loss | `2.31543493` | `2.31939006` |
| 最大 loss | `10.84558201` | `11.12244701` |
| Step 1 loss 绝对差 | - | `4.8e-7` |
| Step 300 loss 绝对差 | - | `0.30019379` |
| 300-step 平均绝对 loss 差 | - | `0.03310791` |
| 最大绝对 loss 差 | - | `0.38271522`，step 100 |
| RMSE | - | `0.06556428` |
| Pearson correlation | - | `0.99916260` |
| 20-step moving average MAE | - | `0.01065584` |
| 20-step moving average Pearson | - | `0.99993332` |
| 最大 LR 绝对差 | - | `4.3e-9` |

补充验证：

- PaddleFormers 和 ms-swift 均有连续的 step 1–300 记录；
- 双方所有 loss 和学习率均为有限数，无 NaN/Inf；
- 双方 step 300 学习率均为 `0.0`；
- PaddleFormers 日志包含 `Training completed.`；
- ms-swift 日志完成至 `300/300`；
- 前两步 input IDs 与 labels 的 SHA-256 已核验一致；
- 完整 300 步未输出双方逐 step batch hash，因此不声明全部 300 步 batch hash 均已验证一致。
### 5. 模型单测

新增：

- `tests/transformers/moondream2/test_modeling.py`

包含 13 个显式 Moondream2 测试方法，覆盖：

- 组合配置的通用配置测试；
- 模型构造和公共接口；
- 视觉编码器输出 shape；
- 非法 crop 尺寸检查；
- 纯文本前向；
- GQA 配置；
- 图文多模态前向；
- BOS 和图像 token 前缀长度；
- 2D/4D attention mask 等价性；
- 图像前缀和 `-100` labels 的 loss 路径；
- `return_dict=False` 返回结构；
- KV cache 增量解码与完整前向最后一个 token logits 一致性；
- generation prefill 后移除图像输入；
- 纯文本和多模态 greedy generation；
- 本地 `save_pretrained` / `from_pretrained` round trip；
- 保存加载前后多模态 logits 一致性。

单测中的 Paddle 内部数值一致性：

- 2D/4D attention mask logits：`atol=1e-6, rtol=1e-6`；
- 保存加载前后 logits：`atol=1e-6, rtol=1e-6`；
- KV cache 与完整解码最后一个 token logits：`atol=1e-5, rtol=1e-5`。

### CODE REVIEW

File: paddleformers/transformers/moondream2/processor.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        image_inputs = self.image_processor(images, **output_kwargs["images_kwargs"]) if images is not None else {}
+        if task is not None:
+            prompt = self.build_task_prompt(task, text=text, length=task_length)
+            if isinstance(prompt, tuple):
+                prefix, question, suffix = prompt
+                tokenized_question = self.tokenizer(question, add_special_tokens=False)["input_ids"]
+                if tokenized_question and isinstance(tokenized_question[0], list):
+                    tokenized_question = tokenized_question[0]
+                input_ids = [prefix + tokenized_question + suffix]
+                text_inputs = {"input_ids": input_ids, "attention_mask": [[1] * len(input_ids[0])]}
```
Comment (risemeup1111) [inline paddleformers/transformers/moondream2/processor.py]: P2 **优先级：P2**

**处理要求：请针对该评论进行回复（同意并已修改请回复 Done，不同意请说明理由）。**

`task` 分支会先把 `text` 交给 `build_task_prompt`，而非 caption 任务里又会执行 `str(text)`。当调用方用 `images=[...]` 和 `text=["q1", "q2"]` 批量做 query/detect/point 时，这里会把整个列表序列化成一个问题，并只返回 batch size=1 的 `input_ids`；但 `image_processor` 已经返回 batch size=N 的 `pixel_values`，后续模型拼接图像前缀和文本 embedding 时 batch 维度会不一致。请明确支持 batch task prompt，或在 `task is not None` 时显式拒绝列表输入，避免静默构造错误样本。

可参考的修复形态：

```python
if isinstance(text, list) and text and all(isinstance(item, str) for item in text):
    questions = text
else:
    questions = [text]

batched_input_ids = []
for question in questions:
    prefix, question, suffix = self.build_task_prompt(task, text=question, length=task_length)
    question_ids = self.tokenizer(question, add_special_tokens=False)["input_ids"]
    if question_ids and isinstance(question_ids[0], list):
        question_ids = question_ids[0]
    batched_input_ids.append(prefix + question_ids + suffix)

text_inputs = {
    "input_ids": batched_input_ids,
    "attention_mask": [[1] * len(ids) for ids in batched_input_ids],
}
```

**Review by risemeup1111 — COMMENTED** [review]
已完成首轮审阅。发现一个非阻塞的 Processor batch 语义问题，细节已放在行级评论里。

- P3 **优先级：P3** 非行级：PR 范围/描述不一致。当前标题和描述聚焦 Moondream2，但 diff 同时新增了 GLM4 模型、配置、测试及 Auto 注册。请将 GLM4 拆到独立 PR，或至少在标题/描述中明确本 PR 同时引入 GLM4，并补充对应测试与风险边界。

**处理要求：请针对该评论进行回复（同意并已修改请回复 Done，不同意请说明理由）。**

另外，GitHub 当前 mergeability 显示为 `CONFLICTING/DIRTY`；合入前需要同步 `develop` 并解决冲突。`gh-llm pr conflict-files` 本次未能定位到具体冲突文件。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>
